### DISCUSSION TITLE
2024-01-05 - Avoiding single points of failure - Chewing the Fat 📈

### DISCUSSION THREAD

**Discussion by bradystroud** [SSWConsulting/SSW.Rules.Content#7697] [Chewing the Fat]
Hey SSWers,

Lets talk about single points of failure and handovers!

https://www.ssw.com.au/rules/do-you-know-how-to-handover-a-project

**Comment by camillars** [comment]
Avoiding single points of failure is essential to keep the business moving and running smoothly.

**Comment by AlexRothwell** [comment]
Single points of failure are important to work on avoiding but the strategy to mitigate the risk must be considered and proportional

For example, throwing out role based protections in response to finding only a couple of people holding the role may speed things up but could also end in disaster

**Comment by 2fernandez** [comment]
I am super confused with this week's CTF. Are we talking about:
"Project handover" or "Single points of failure" ?

Are we suggesting that lack of "Project handover" can cause "Single points of failure" ?

**Comment by Dhruv-0987** [comment]
The most important thing I learnt was to not mess with Username for one and the importance of secured authentication. Also how returning true instead of false can literally shake the world.

**Comment by TobyChurches** [comment]
Its definitely important to mitigate single points of failure, especially in large/growing organisations like SSW. Its great to see that task responsibilities are being entrusted to multiple employees in SL, It can be a bit hard to get onto the correct person otherwise 🙂

Still for a number of tasks in SL require a single 'final' approver - Maybe this could be sent to multiple people for approval?

**Comment by Calinator444** [comment]
I was shocked that the Google presentation was based on a real-life example. Allowing users to login without passwords is the kind of security flaw that drives a company out of business, if not resulting in a hefty lawsuit. Dropbox must have handled the situation very tactfully to stay afloat afterwards.

**Comment by jimmy1992abc** [comment]
Ensuring that someone is designated to support you during your absence is crucial, and we strongly advocate for establishing this as a standard or rule. 

The rules discussed in today's CTF seem more aligned with transitioning responsibilities when someone permanently leaves a role, emphasizing task handovers. 

However, for a seamless delegation of responsibilities during a holiday, our emphasis should be on task organization, identifying alternative contacts, and establishing emergency communication etc.

**Comment by wicksipedia** [comment]
No mention of the [bus test](https://medium.com/swlh/the-bus-test-8c5e2a2d7507) which was surprising

**Reply by jaydenalchin** [reply]
+1 plus the [key person risk](https://www.jadeworld.com/resources/blog/technology-key-person-risk)

**Comment by YalalCakirSSW** [comment]
A project shouldn't just rely on a couple of people, and the system should operate seamlessly even when key team members are on holidays. To be able to achieve this, a good handover is crucial. Preventing a single point of failure by ensuring that essential information, responsibilities, and processes are effectively transferred between team members.

**Comment by jeoffreyfischer** [comment]
Eliminating single points of failure is crucial for maintaining uninterrupted and smooth business operations.

**Comment by tkapa** [comment]
There's been a couple of times - not even just while I've been at SSW that redundancy or handover processes would have been important. 

These kinds of things are hard to spot without persistent analysis, or until it's too late.

**Comment by chrisschultzssw** [comment]
Important stuff - and good to see SysAdmin related examples :).

**Comment by vladislav-kir** [comment]
Preventing single points of failure is crucial! 

It's great to know SSW's template for project handover :)

**Comment by CoffeeBugHub** [comment]
Handover is important, and it is always a good practice to perform a proper handover when you are going to be on leave, ensuring that you do not leave the team or other people in a difficult situation. 

Even if you are working on certain PBIs and are suddenly assigned to client work, it is always good practice to document what you have done and what remains. This way, the people taking over do not have to chase you for a handover.

**Comment by cvionnet** [comment]
Nice to have the 8 steps for the project hand-over

**Comment by christoment** [comment]
Mitigating the risk of single point of failure is critical - and I think this is very valuable.

But of course, with every mitigation comes with the $$ - so as a consultant it's also important to find the appropriate mitigation strategy for the given risk 😄

**Reply by GordonBeeming** [reply]
👆 this... it's easy to over-engineer systems and processes to the point where they become hard to work with, and more time and money are wasted from using the process than a little 'downtime' (both in systems and process) that could be caused by something going wrong. 
It's important to understand the appetite for risk when designing anything

**Comment by joshbermanssw** [comment]
Proper handovers can mitigate so many issues down the track. The 8 rules delve into perhaps unthought of tasks -> checking all known passwords work is crucial!

**Comment by ncn-ssw** [comment]
Single points of failure are sometimes unavoidable within the reasonable constraints of a project, but their risk should always be actively assessed and managed.

**Comment by drwharris** [comment]
Single points of failure are an inevitability. What's important is how you respond to them. Do you just complain, pointing out the "stupidity" of things or do you jump on the front foot and start to find a solution to the problem.

**Comment by calumjs** [comment]
There is always some counter party risk if you are not in sole possession of the private keys

**Comment by amankumarrr** [comment]
Decent handover can help address certain issues, but it doesn't eliminate the risk of a single point of failure. It's essential to evaluate risks and establish failover strategies.

**Comment by KahaMason** [comment]
Identifying and mitigating single points of failures are definitely crucial for ensuring smooth operation. Unfortunately there can be a lot of times where this can go unnoticed until something goes wrong or something is blocked.

It was good to see that SSW has a template for project handover. I've had the experience myself where sometimes tasks will fall into limbo due to little to no handover process.

**Comment by UlyssesMaclaren** [comment]
Reminds me a lot of scaling issues in general. SPOF is often also a bottleneck for growth.

Having something like the Induction Masters where you can keep going down the list till you find someone available is one solution

**Comment by ricksu978** [comment]
It's clear that avoiding single points of failure is crucial for smooth operations. Handovers play a key role in this, ensuring knowledge and responsibilities are effectively transferred.

**Comment by steven0x51** [comment]
There's always single point of failure in a system, what matters is to balance the cost and risk

**Comment by MattParkerDev** [comment]
It is important to share knowledge and reduce SPF, however sometimes it is unavoidable and can only be mitigated.

**Comment by SamBWagner** [comment]
Knowledge sharing is key and having choke points is something to be avoided at all costs if possible.

**Comment by bettybondoc** [comment]
By doing a proper handover, we make sure everything runs smoothly without any hiccups or mistakes, making transitions a breeze and keeping our high standards intact. But I also believe it's really important to keep everyone in the loop about what you're doing all the time (eg. saving files in SharePoint, keeping your PBIs up to date and very detailed), not just when you're heading out for a break. This way, if something comes up out of the blue, anyone can step in and keep the ball rolling without any fuss.

**Comment by AttackOnMorty** [comment]
Pair programming helps.

**Reply by jimmidier** [reply]
sounds like something I should print on my T shirt😄

**Comment by AntPolkanov** [comment]
Handover of login details is a pain. Ideally, there shouldn't be any handover of credentials as they should be stored in password managers.

**Comment by JackDuan123** [comment]
Interesting thought:
What if the whole world we live in is on a single point of failure, if one thing fails then the whole world will collapse?
Or it's completely virtual, and if the computer restart by the creator of us, then becuase of not being able to backup properly, someone will just disappear from us and we don't even know?

**Reply by calumjs** [reply]
🤯

**Reply by jimmidier** [reply]
we live on math and it's the single point of failure in a sense. As the saying goes, by 4 percent alteration in the fine structure constant there may not be carbon and oxygen elements

**Comment by JackDevAU** [comment]
I think we should avoid single points of failure as much as possible but sometimes it just isn't possible. We need backup plans to the backup plans!!  
Good to think about though and should be something that people are thinking about from time to time and reporting if found.

**Comment by JackDuan123** [comment]
Serious one after joke:
A proper documentation is always useful when we handover project or when we check the structure we can see what's going wrong.

There is no fool-proof process or orgnization that can erase human mistakes, so being able to have a test team to find possible failures is also important.

**Comment by synnoveisabel** [comment]
Good CTF, it can be a pain when people go on leave, avoiding a single point of failure ensures that work continues smoothly without depending on just one person.

**Comment by PennyWalker** [comment]
Initiating the co-auth & Masters system has helped us mitigated my biggest point of failure which is not being able to reach Adam for an approval. It's great to have the option to activate the phone tree and get unblocked

**Comment by jerryluossw** [comment]
I like  8 steps for a good handover very much. 
I totlly agree avoiding single points of failure is so important, but I think we can't have redundancy in everything.
Redundancy means higher costs (Much more expensive than single points), we need to think about the balance between redundancy and cost.

**Comment by Aibono1225** [comment]
It's good to learn from Google's example about how serious a single point of failure can be.

**Comment by jimmidier** [comment]
IMO dealing with single points of failure in a sociological context can be generalized in 2 categories,
- Readiness-based business - things like this rely on information propagation thus can be accelerated by reducing duplication cost, typically by digitization
- Completion-based business - things like this rely on a distributed system so a consensus mechanism helps in this case

It's like a metaphor of a software system except the latter has the natural advantage of being natively digitized.

**Comment by Levijj22** [comment]
Great CTF!
Loved the video! Really makes you think about how one mistake/decision can collapse an entire system.
Scientists are saying we could be facing a complete internet outage (for up to a few months) in 2024 due to solar flares, what's our backup then?  👀

**Comment by babakamyljanovssw** [comment]
Loved the rule for handover the project. Following the steps and template from the rule definitely helps team's work to be smooth. Project handover can mitigate Single Point of Failure, but there is a always chance for SPF, and it is crucial to establish strategies for it.

**Comment by tino-liu** [comment]
Great Rule! Smooth handovers are crucial to a company's reputation as well.

**Comment by sethdaily** [comment]
Very important. This should be in everyone's mind when working on processes

**Comment by jakebayliss** [comment]
My project is especially key person dependent (on the client side). When they are away we have to push tasks back until they return. It's not great. But it seems like this is something SSW is good at as we have always have many devs to a project. Good documentation is key

**Comment by bahjatalaadelssw** [comment]
Single point failure occurs with poor documentation which is a result of complex processes not allowing for collection of information to maintain a chain of accountability

**Comment by Freego1783** [comment]
Knowing how to handover a project is really important. I witnessed the acquisition of one company by another and the transfer of a project from one development team to the other. No effort was made, and it was very complicated for the buyers to continue the project.

**Comment by stephanfakossw** [comment]
It's super important to avoid systems and processes failing because of lack of redundancies .

**Comment by piers-sinclair** [comment]
Imo, cross approvals are the golden ticket 🎫

**Comment by danielmackay** [comment]
single points of failure are important to keep in mind when dealing with both systems, people, and processes.  This CTF is a good reminder to keep any eye out and address any that we come across.

**Comment by Marxoz** [comment]
I think the golden rule should be that a process / approval does not sit with only one person. Having 2 tiers is probably also a good idea. One tier that does the approvals / ownership, and another that is available only for emergencies. Having this documented is also important, so that you know who to contact, but also so that the person who is going on leave knows what he needs to handover.

**Comment by BrookJeynes** [comment]
It's important to minimise the amount of single points of failure on any given projects. Having good handover docs and easy handover processes can help with this.
